packages/devtools_app/test/memory/shared/heap/heap_test.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:devtools_app/src/screens/memory/shared/heap/heap.dart';
+import 'package:devtools_app/src/screens/memory/shared/heap/model.dart';
+import 'package:devtools_app/src/screens/memory/shared/heap/spanning_tree.dart';
+import 'package:devtools_app/src/screens/memory/shared/primitives/class_name.dart';
+
+import 'package:flutter_test/flutter_test.dart';
+
+class _ClassSizeTest {
+  _ClassSizeTest({
+    required this.name,
+    required this.heap,
+    required this.expectedClassARetainedSize,
+  }) : assert(_assertHeapIndexIsCode(heap)) {
+    buildSpanningTree(heap);
+  }
+
+  final AdaptedHeapData heap;
+  final String name;
+  final int expectedClassARetainedSize;
+}
+
+final _root = HeapClassName(className: 'Root', library: 'l');
+final _classA = HeapClassName(className: 'A', library: 'l');
+final _classB = HeapClassName(className: 'B', library: 'l');
+
+final _classSizeTests = <_ClassSizeTest>[
+  _ClassSizeTest(
+    name: 'separate',
+    heap: AdaptedHeapData(
+      [
+        _createOneByteObject(0, [1, 2, 3], _root),
+        _createOneByteObject(1, [], _classA),
+        _createOneByteObject(2, [], _classA),
+        _createOneByteObject(3, [], _classA),
+      ],
+      rootIndex: 0,
+    ),
+    expectedClassARetainedSize: 3,
+  ),
+  _ClassSizeTest(
+    name: 'linked',
+    heap: AdaptedHeapData(
+      [
+        _createOneByteObject(0, [1], _root),
+        _createOneByteObject(1, [2], _classA),
+        _createOneByteObject(2, [3], _classA),
+        _createOneByteObject(3, [], _classA),
+      ],
+      rootIndex: 0,
+    ),
+    expectedClassARetainedSize: 3,
+  ),
+  _ClassSizeTest(
+    name: 'full graph',
+    heap: AdaptedHeapData(
+      [
+        _createOneByteObject(0, [1], _root),
+        _createOneByteObject(1, [2, 3], _classA),
+        _createOneByteObject(2, [3, 1], _classA),
+        _createOneByteObject(3, [1, 2], _classA),
+      ],
+      rootIndex: 0,
+    ),
+    expectedClassARetainedSize: 3,
+  ),
+  _ClassSizeTest(
+    name: 'with global B',
+    heap: AdaptedHeapData(
+      [
+        _createOneByteObject(0, [1], _root),
+        _createOneByteObject(1, [2, 4], _classA),
+        _createOneByteObject(2, [3, 4], _classA),
+        _createOneByteObject(3, [4], _classA),
+        _createOneByteObject(4, [], _classB),
+      ],
+      rootIndex: 0,
+    ),
+    expectedClassARetainedSize: 4,
+  ),
+];
+
+void main() {
+  for (final t in _classSizeTests) {
+    test(
```

Comment:
In general avoid dynamically creating test cases like this. It makes it hard for IDE tooling. 
@natebosch would you be ok with a lint saying not to nest `test` in a for loop?
Instead have a single test case and in the test case make sure the expect clauses make it unambiguous which class name the failure is about. For example by augmenting the output returned for a failure or by having multiple expect statements one for each class name in the heap you are interested in.